File: Upupu/UploadViewController.m
//
//  UploadViewControllerDelegate.m
//  Upupu
//
//  Created by David Ott on 11/23/11.
//  Copyright 2011 Xcoo, Inc. All rights reserved.
//

#import "UploadViewController.h"
#import "UpupuAppDelegate.h"

#import "WebDAVUploader.h"
#import "DropboxUploader.h"

#import "Settings.h"

#import "ImageUtil.h"
#import "HUDUtil.h"
#import "AlertUtil.h"

@interface UploadViewController () <MBProgressHUDDelegate, UITextFieldDelegate> {
    MBProgressHUD *_hud;
}

- (IBAction)retake:(id)sender;
- (IBAction)upload:(id)sender;
- (IBAction)setting:(id)sender;

@property (nonatomic, retain) IBOutlet UITextField *nameField;
@property (nonatomic, retain) IBOutlet UIImageView *imageView;
@property (nonatomic, retain) IBOutlet UIBarItem *retakeButton;
@property (nonatomic, retain) IBOutlet UIBarItem *uploadButton;
@property (nonatomic, retain) IBOutlet UIBarItem *settingButton;

@end

@implementation UploadViewController

@synthesize nameField = _nameField, imageView = _imageView;
@synthesize retakeButton = _retakeButton, uploadButton = _uploadButton, settingButton = _settingButton;
@synthesize image = _image;
@synthesize delegate = _delegate;
@synthesize savePhotoAlbum = _savePhotoAlbum;

#pragma mark - Class Method -

+ (NSString *)makeFilename
{
    NSDate *dateCreated = [[[NSDate alloc] init] autorelease];
    NSDateFormatter *dateFormatter = [[[NSDateFormatter alloc] init] autorelease];
    [dateFormatter setDateFormat:@"yyyyMMdd_HHmmss"];
    return [dateFormatter stringFromDate:dateCreated];
}

#pragma mark - Actions -

-(IBAction) retake:(id)sender
{
    _nameField.text = @"";
    if( _delegate != nil && [_delegate respondsToSelector:@selector(uploadViewControllerDidReturn:)] ) {
        [_delegate uploadViewControllerDidReturn:self];
    }
}

- (IBAction) upload:(id)sender
{
    if (![Settings isWebDAVEnabled] && ![Settings isDropboxEnabled]) {
        [AlertUtil showWithTitle:@"Error" andMessage:@"Setup server configuration before uploading"];
        return;
    }
    
    if ([Settings isWebDAVEnabled] &&
        ([Settings webDAVURL] == nil || [[Settings webDAVURL] isEqualToString:@""])) {
        [AlertUtil showWithTitle:@"Error" andMessage:@"Invalid WebDAV server URL"];
        return;
    }
    
    _hud = [[HUDUtil showWithText:@"Uploading" forView:self.navigationController.view whileExecuting:@selector(lauchUpload) onTarget:self] retain];
}

- (IBAction) setting:(id)sender
{
    if( _delegate != nil && [_delegate respondsToSelector:@selector(uploadViewControllerDidSetup:)] ) {
        [_delegate uploadViewControllerDidSetup:self];
    }
}

#pragma mark - Take Picture -

- (void) showFailed
{
    _hud.customView = [[[UIImageView alloc] initWithImage:[UIImage imageNamed:@"failure_icon"]] autorelease];
    _hud.mode = MBProgressHUDModeCustomView;
    _hud.labelText = @"Failed";
    _hud.detailsLabelText = @"";
}

- (void) showSucceeded
{
    _hud.customView = [[[UIImageView alloc] initWithImage:[UIImage imageNamed:@"success_icon"]] autorelease];
    _hud.mode = MBProgressHUDModeCustomView;
    _hud.labelText = @"Succeded";
    _hud.detailsLabelText = @"";
}

- (void) lauchUpload
{
    UIImage *image = nil;
    
    // resize image
    switch ([Settings photoResolution]) {
        case 0:
        default:
            image = _image;
            break;
        case 1:
            image = [ImageUtil scaleImage:_image withSize:CGSizeMake(1600, 1200)];
            break;
        case 2:
            image = [ImageUtil scaleImage:_image withSize:CGSizeMake(800, 600)];
            break;
    }
    
    // convert to JPG data
    CGFloat quality;
    switch ([Settings photoQuality]) {
        case 0:
        default:
            quality = 1.0f; // High
            break;
        case 1:
            quality = 0.6f; // Mideum
            break;
        case 2:
            quality = 0.2f; // Low
            break;
    }
    
    if (_savePhotoAlbum && [Settings photoSaveToAlbum])
        UIImageWriteToSavedPhotosAlbum(image, nil, nil, nil);
    
    NSData *imageData = UIImageJPEGRepresentation(image, quality);
    
    NSString *filename = [_nameField text];

    // WebDAV
    if ( [Settings isWebDAVEnabled] ) {
        _hud.detailsLabelText = @"WebDAV";
        WebDAVUploader *uploader = [[WebDAVUploader alloc] initWithName:filename imageData:imageData];
        [uploader upload];
        if ( !uploader.success ) {
            dispatch_sync(dispatch_get_main_queue(), ^{
                [self showFailed];
            });
            sleep(1);
            return;
        }
        [uploader release];
    }
    
    // Dropbox
    if ( [Settings isDropboxEnabled] ) {
        _hud.detailsLabelText = @"Dropbox";
        DropboxUploader *uploader = [DropboxUploader sharedInstance];
        [uploader uploadWithName:filename imageData:imageData];
        if ( !uploader.success ) {
            [self showFailed];
            sleep(1);
            return;
        }
    }
    
    // all succeeded
    dispatch_sync(dispatch_get_main_queue(), ^{
        [self showSucceeded];
    });
    sleep(1);

    // finish
    _nameField.text = @"";
    if( _delegate != nil && [_delegate respondsToSelector:@selector(uploadViewControllerDidFinished:)] ) {
        [_delegate uploadViewControllerDidFinished:self];
    }
}

#pragma mark - View lifecycle -

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (void)viewWillAppear:(BOOL)animated
{
    _imageView.image = _image;
    
    if (!_image) {
        _nameField.enabled = NO;
        _uploadButton.enabled = NO;
    } else {
        _nameField.enabled = YES;
        _uploadButton.enabled = YES;
    }
    
    if (_nameField.text.length == 0) {
        _nameField.text = [UploadViewController makeFilename];
    }
    
    [super viewWillAppear:animated];
}

- (BOOL) shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

- (BOOL)shouldAutorotate
{
	return YES;
}

- (NSUInteger)supportedInterfaceOrientations
{
	return UIInterfaceOrientationMaskPortrait;
}

- (void) viewDidUnload
{
    SAFE_RELEASE(_image)
    
    SAFE_RELEASE(_nameField)
    SAFE_RELEASE(_imageView)
    SAFE_RELEASE(_retakeButton)
    SAFE_RELEASE(_uploadButton)
    SAFE_RELEASE(_settingButton)

    SAFE_RELEASE(_hud)
    
    [super viewDidUnload];
    
    Logging(@"deallocated");
}

- (void)dealloc
{
    SAFE_RELEASE(_image)
    
    SAFE_RELEASE(_nameField)
    SAFE_RELEASE(_imageView)
    SAFE_RELEASE(_retakeButton)
    SAFE_RELEASE(_uploadButton)
    SAFE_RELEASE(_settingButton)
    
    SAFE_RELEASE(_hud)
    
    [super dealloc];
    
    Logging(@"deallocated");
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

#pragma mark - TextFieldDelegate -

- (BOOL)textFieldShouldReturn:(UITextField *)textField 
{
    [textField resignFirstResponder]; 
    return YES;
}

@end

File: Upupu/Controllers/UploadViewController.swift
//
//  UploadViewController.swift
//  Upupu
//
//  Created by Toshiki Takeuchi on 8/29/16.
//  Copyright © 2016 Xcoo, Inc. All rights reserved.
//  See LISENCE for Upupu's licensing information.
//

import UIKit

import Alamofire
import MBProgressHUD

protocol UploadViewControllerDelegate: class {

    func uploadViewControllerDidReturn(_ uploadViewController: UploadViewController)
    func uploadViewControllerDidFinished(_ uploadViewController: UploadViewController)
    func uploadViewControllerDidSetup(_ uploadViewController: UploadViewController)

}

class UploadViewController: UIViewController, MBProgressHUDDelegate, UITextFieldDelegate {

    weak var delegate: UploadViewControllerDelegate?

    private var uploadView: UploadView!

    var image: UIImage?
    var shouldSavePhotoAlbum = true

    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func loadView() {
        uploadView = UploadView()
        view = uploadView
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        uploadView.retakeButton.action = #selector(retakeButtonTapped)
        uploadView.uploadButton.action = #selector(uploadButtonTapped)
        uploadView.settingsButton.action = #selector(settingsButtonTapped)

        uploadView.nameTextField.delegate = self
    }

    override func viewWillAppear(_ animated: Bool) {
        uploadView.imageView.image = image

        uploadView.nameTextField.isEnabled = image != nil
        uploadView.uploadButton.isEnabled = image != nil

        if let text = uploadView.nameTextField.text {
            if text.isEmpty {
                uploadView.nameTextField.fileStem = makeFilename()
            }
        } else {
            uploadView.nameTextField.fileStem = makeFilename()
        }

        super.viewWillAppear(animated)
    }

    override var shouldAutorotate: Bool {
        return UIDevice.current.orientation == .portrait
    }

    override var prefersStatusBarHidden: Bool {
        return false
    }

    override var preferredStatusBarUpdateAnimation: UIStatusBarAnimation {
        return .fade
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        view.endEditing(true)
    }

    private func makeFilename() -> String {
        return DateHelper.dateTimeString()
    }

    // MARK: - Action

    @objc private func retakeButtonTapped(_ sender: UIBarItem) {
        uploadView.nameTextField.text = ""
        delegate?.uploadViewControllerDidReturn(self)
    }

    @objc private func uploadButtonTapped(_ sender: UIBarItem) {
        guard Settings.webDAVEnabled || Settings.dropboxEnabled else {
            UIAlertController.showSimpleErrorAlertIn(navigationController,
                                                     error: UPError.settingsNotSetUp)
            return
        }

        // Checking network connection
        guard let net = NetworkReachabilityManager() else {
            UIAlertController.showSimpleErrorAlertIn(navigationController,
                                                     error: .networkUnreachable)
            return
        }
        net.startListening()
        let isReachable = net.isReachable
        net.stopListening()
        guard isReachable else {
            UIAlertController.showSimpleErrorAlertIn(navigationController,
                                                     error: .networkUnreachable)
            return
        }

        let hud = MBProgressHUD.showAdded(to: view, animated: true)
        DispatchQueue.global(qos: .background).async {[weak self] in
            self?.launchUpload(hud)
        }
    }

    @objc private func settingsButtonTapped(_ sender: UIBarItem) {
        delegate?.uploadViewControllerDidSetup(self)
    }

    // MARK: - Picture processing

    private func scaleImage(_ image: UIImage?) -> UIImage? {
        switch Settings.photoResolution {
        case .original:
            return image
        case .medium:
            return image?.scaledImage(CGSize.init(width: 1600, height: 1200))
        case .small:
            return image?.scaledImage(CGSize.init(width: 800, height: 600))
        }
    }

    private func imageData(_ image: UIImage) -> Data? {
        let quality: Float
        switch Settings.photoQuality {
        case .high:
            quality = 1.0
        case .medium:
            quality = 0.6
        case .low:
            quality = 0.2
        }
        return UIImageJPEGRepresentation(image, CGFloat(quality))
    }

    private func showFailed(_ hud: MBProgressHUD?, message: String? = nil) {
        if let hud = hud {
            hud.customView = UIImageView(image: UIImage(named: "Upload/Failure"))
            hud.mode = .customView
            hud.label.text = "Failed"
            hud.detailsLabel.text = message
        }
    }

    private func showSucceeded(_ hud: MBProgressHUD?) {
        if let hud = hud {
            hud.customView = UIImageView(image: UIImage(named: "Upload/Success"))
            hud.mode = .customView
            hud.label.text = "Succeeded"
            hud.detailsLabel.text = ""
        }
    }

    private func execUpload<T: Uploadable>(_ uploader: T, filename: String, imageData: Data,
                                           hud: MBProgressHUD?) {
        uploader.upload(filename, data: imageData) { (error) in
            guard error == nil else {
                DispatchQueue.main.async {[weak self] in
                    self?.showFailed(hud, message: error?.description)
                }

                DispatchQueue.main.asyncAfter(deadline: .now() + 2) {
                    hud?.hide(animated: true)
                }

                return
            }

            DispatchQueue.main.async {[weak self] in
                self?.showSucceeded(hud)
            }

            DispatchQueue.main.asyncAfter(deadline: .now() + 1) {[weak self] in
                hud?.hide(animated: true)
                self?.uploadView.nameTextField.text = ""
                if let self_ = self {
                    self_.delegate?.uploadViewControllerDidFinished(self_)
                }
            }
        }
    }

    private func launchUpload(_ hud: MBProgressHUD?) {
        guard let image = scaleImage(self.image) else {
            return
        }

        // Save to album
        if shouldSavePhotoAlbum && Settings.shouldSavePhoto {
            UIImageWriteToSavedPhotosAlbum(image, nil, nil, nil)
        }

        // Upload to ...
        if let imageData = imageData(image) {
            let fileStem = uploadView.nameTextField.fileStem
            let filename: String
            if fileStem == nil || fileStem!.isEmpty {
                filename = "\(Uploader.fileStem()).jpg"
            } else {
                filename = "\(fileStem!).jpg"
            }

            // WebDAV
            if Settings.webDAVEnabled {
                DispatchQueue.main.sync(execute: {
                    hud?.detailsLabel.text = "WebDAV"
                    })
                execUpload(WebDAVUploader(), filename: filename, imageData: imageData, hud: hud)
            }

            // Dropbox
            if Settings.dropboxEnabled {
                DispatchQueue.main.sync(execute: {
                    hud?.detailsLabel.text = "Dropbox"
                    })
                execUpload(DropboxUploader(), filename: filename, imageData: imageData, hud: hud)
            }
        }
    }

    // MARK: - TextFieldDelegate

    func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
        if let filenameTextField = textField as? FilenameTextField {
            filenameTextField.extentionHidden = true
        }
        return true
    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return true
    }

    func textFieldDidEndEditing(_ textField: UITextField) {
        if let filenameTextField = textField as? FilenameTextField {
            filenameTextField.extentionHidden = false
        }
    }

}
